// qier-player/dist/es/main/modules/loading/index.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { CLASS_PREFIX, EVENT } from '../../constants';
import { addDispose } from '../../utils/dispose';
import { addChildren, addClass, hide, show } from '../../utils/dom';
import { DomNode } from '../../utils/domNode';
import { repeatStr } from '../../utils/freUse';
var Loading = /** @class */ (function (_super) {
    __extends(Loading, _super);
    function Loading(container, player) {
        var _this = _super.call(this, container, '.loading') || this;
        _this.player = player;
        _this.startWaitingTime = 0;
        _this._checkCanplay = function () {
            if (_this.startWaitingTime !== _this.player.currentTime) {
                _this.hide();
                _this.clearTimeout();
                _this.player.off(EVENT.TIME_UPDATE, _this._checkCanplay);
            }
        };
        _this.clearTimeout = function () {
            _this.showTimer && clearTimeout(_this.showTimer);
            _this.showTimer = null;
        };
        _this.show = function () {
            show(_this.el);
            _this.player.emit(EVENT.LOADING_SHOW);
        };
        _this.hide = function () {
            hide(_this.el);
            _this.player.emit(EVENT.LOADING_HIDE);
        };
        _this.hide();
        if (player.options.loadingOptions.spinner) {
            addChildren(_this.el, [player.options.loadingOptions.spinner]);
        }
        else {
            addChildren(_this.el, _this.getSpinner());
        }
        addDispose(_this, player.on(EVENT.CANPLAY, _this.hide));
        addDispose(_this, player.on(EVENT.WAITING, function () {
            if (!_this.player.currentTime)
                return;
            _this.tryShow();
        }));
        addDispose(_this, player.on(EVENT.STALLED, function () {
            var curTime = _this.player.currentTime;
            if (!curTime)
                return;
            var shouldShow = true;
            _this.player.eachBuffer(function (start, end) {
                if (start <= curTime && end > curTime) {
                    shouldShow = false;
                    return true;
                }
            });
            if (shouldShow)
                _this.tryShow();
        }));
        return _this;
    }
    Loading.prototype.getSpinner = function () {
        var type = this.player.options.loadingOptions.type;
        if (type === 'wave') {
            addClass(this.el, 'loading_wave');
            return repeatStr('<div></div>', 5);
        }
        if (type === 'circle') {
            addClass(this.el, 'loading_circle');
            return "<div>\n            <div class=\"" + CLASS_PREFIX + "_loading_circle_container\">\n              <div class=\"" + CLASS_PREFIX + "_loading_circle_rotator\">\n                <div class=\"" + CLASS_PREFIX + "_loading_circle_left\">\n                  <div class=\"" + CLASS_PREFIX + "_loading_circle_circle\"></div>\n                </div>\n                <div class=\"" + CLASS_PREFIX + "_loading_circle_right\">\n                  <div class=\"" + CLASS_PREFIX + "_loading_circle_circle\"></div>\n                </div>\n              </div>\n            </div>\n          </div>";
        }
    };
    Object.defineProperty(Loading.prototype, "isActive", {
        get: function () {
            return this.el.style.display !== 'none';
        },
        enumerable: false,
        configurable: true
    });
    Loading.prototype.checkCanplay = function () {
        this.startWaitingTime = this.player.currentTime;
        this.player.off(EVENT.TIME_UPDATE, this._checkCanplay);
        this.player.on(EVENT.TIME_UPDATE, this._checkCanplay);
    };
    Loading.prototype.tryShow = function () {
        this.checkCanplay();
        this.clearTimeout();
        this.showTimer = setTimeout(this.show, 300);
    };
    return Loading;
}(DomNode));
export { Loading };
