// qier-player/dist/es/main/components/tooltip.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { DomNode } from '../utils/domNode';
import { addClass, removeClass, hide, show } from '../utils/dom';
var classLeft = 'tooltip-left';
var classRight = 'tooltip-right';
var classBottom = 'tooltip-bottom';
var Tooltip = /** @class */ (function (_super) {
    __extends(Tooltip, _super);
    function Tooltip(container, html) {
        var _this = this;
        addClass(container, 'tooltip');
        _this = _super.call(this, container, 'div.tooltip_content') || this;
        if (html)
            _this.html = html;
        return _this;
    }
    Object.defineProperty(Tooltip.prototype, "html", {
        get: function () {
            return this.el.innerHTML;
        },
        set: function (s) {
            this.el.innerHTML = s;
        },
        enumerable: false,
        configurable: true
    });
    Tooltip.prototype.resetPos = function () {
        removeClass(this.el, classLeft);
        removeClass(this.el, classRight);
        removeClass(this.el, classBottom);
    };
    Tooltip.prototype.setBottom = function () {
        addClass(this.el, classBottom);
    };
    Tooltip.prototype.setLeft = function () {
        addClass(this.el, classLeft);
    };
    Tooltip.prototype.setRight = function () {
        addClass(this.el, classRight);
    };
    Tooltip.prototype.hide = function () {
        hide(this.el);
    };
    Tooltip.prototype.show = function () {
        show(this.el);
    };
    return Tooltip;
}(DomNode));
export { Tooltip };
