// qier-player/dist/es/main/utils/dom.js
import { isString, isHTMLElement } from './is';
import { CLASS_PREFIX } from '../constants';
// e.g. 'div#id.className' to get 'div', 'id', '.className'
var SELECTOR_REGEX = /([\w-]+)?(?:#([\w-]+))?((?:\.(?:[\w-]+))*)/;
export function createEle(tag, attrs, children, classPrefix) {
    if (classPrefix === void 0) { classPrefix = CLASS_PREFIX; }
    var diviTag = classPrefix === '' ? '' : '_';
    var match = [];
    if (tag)
        match = SELECTOR_REGEX.exec(tag) || [];
    var ele = document.createElement(match[1] || 'div');
    if (match[2])
        ele.id = match[2];
    if (match[3])
        ele.className = match[3].replace(/\./g, " " + classPrefix + diviTag).trim();
    if (attrs) {
        Object.keys(attrs).forEach(function (attr) {
            var value = attrs[attr];
            if (value === undefined)
                return;
            if (/^on\w+$/.test(attr)) {
                ele[attr] = value;
            }
            else {
                ele.setAttribute(attr, value);
            }
        });
    }
    if (children) {
        if (isString(children)) {
            ele.innerHTML = children;
        }
        else {
            children.forEach(function (c) { return ele.appendChild(c); });
        }
    }
    return ele;
}
export function addChildren(container, child) {
    if (!child)
        return;
    if (isString(child)) {
        container.innerHTML = child;
    }
    else {
        child.forEach(function (c) { return container.appendChild(c); });
    }
}
export function getEle(el) {
    if (!el)
        return null;
    if (isString(el))
        return document.querySelector(el);
    if (isHTMLElement(el))
        return el;
    throw new Error('The container element you are currently passing in is not an HTML element');
}
export function removeEle(el) {
    if (!el)
        return;
    if (el.remove) {
        el.remove();
    }
    else if (el.parentNode) {
        el.parentNode.removeChild(el);
    }
}
export function removeAllChildNode(el) {
    if (!el)
        return;
    var childs = el.childNodes;
    for (var i = childs.length - 1; i >= 0; i--) {
        el.removeChild(childs[i]);
    }
}
export function show(node) {
    node.style.display = '';
}
export function hide(node) {
    node.style.display = 'none';
}
export function addClass(dom, cls, prefix) {
    if (cls === void 0) { cls = ''; }
    if (prefix === void 0) { prefix = CLASS_PREFIX; }
    cls = cls.trim();
    if (!cls)
        return dom;
    if (dom.classList) {
        cls.split(' ').forEach(function (c) { return dom.classList.add(prefix + "_" + c); });
    }
    else {
        var oldCls = (dom.className && dom.className.baseVal) || '';
        dom.setAttribute('class', (oldCls ? oldCls + " " : '') +
            cls
                .split(' ')
                .map(function (c) { return prefix + c; })
                .join(' '));
    }
    return dom;
}
export function removeClass(dom, cls, prefix) {
    if (prefix === void 0) { prefix = CLASS_PREFIX; }
    dom.classList.remove(prefix + "_" + cls);
    return dom;
}
export function containClass(dom, cls, prefix) {
    if (prefix === void 0) { prefix = CLASS_PREFIX; }
    return dom.classList.contains(prefix + "_" + cls);
}
export function toggleClass(dom, cls, force, prefix) {
    if (prefix === void 0) { prefix = CLASS_PREFIX; }
    cls = prefix + "_" + cls;
    if (force) {
        dom.classList.add(cls);
        return true;
    }
    if (force === false) {
        dom.classList.remove(cls);
        return true;
    }
    return dom.classList.toggle(cls, force);
}
var svgNS = 'http://www.w3.org/2000/svg';
export function createSvg(cls, d, viewBox) {
    if (viewBox === void 0) { viewBox = '0 0 1024 1024'; }
    var svg = document.createElementNS(svgNS, 'svg');
    svg.setAttribute('viewBox', viewBox);
    if (cls)
        addClass(svg, cls);
    if (d) {
        var path = document.createElementNS(svgNS, 'path');
        path.setAttributeNS(null, 'd', d);
        svg.appendChild(path);
    }
    return svg;
}
export function getEventPath(ev) {
    return ev.path || ev.composedPath();
}
var DomListener = /** @class */ (function () {
    function DomListener(node, type, handler, options) {
        this.node = node;
        this.type = type;
        this.handler = handler;
        this.options = options;
        // 对于 options 的支持暂未判断
        node.addEventListener(type, handler, this.options);
    }
    DomListener.prototype.dispose = function () {
        if (!this.handler)
            return;
        this.node.removeEventListener(this.type, this.handler, this.options);
        this.node = null;
        this.handler = null;
        this.options = null;
    };
    return DomListener;
}());
export { DomListener };
export function getElementSize(dom) {
    var clone = dom.cloneNode(true);
    clone.style.cssText += ';position:absolute;opacity:0';
    clone.removeAttribute('hidden');
    var parent = dom.parentNode || document.body;
    parent.appendChild(clone);
    var rect = clone.getBoundingClientRect();
    parent.removeChild(clone);
    return {
        width: rect.width,
        height: rect.height,
    };
}
