// qier-player/dist/es/main/utils/drag.js
var Drag = /** @class */ (function () {
    function Drag(el, start, move, end) {
        var _this = this;
        this.el = el;
        this.pending = false;
        this.downHandler = function (ev) {
            ev.preventDefault();
            _this.el.setPointerCapture(ev.pointerId);
            _this.el.addEventListener('pointermove', _this.moveHandler);
            _this.start(ev);
        };
        this.upHandler = function (ev) {
            ev.preventDefault();
            _this.el.releasePointerCapture(ev.pointerId);
            _this.el.removeEventListener('pointermove', _this.moveHandler);
            if (_this.end)
                _this.end(ev);
        };
        this.moveHandler = function (ev) {
            ev.preventDefault();
            _this.lastEv = ev;
            if (_this.pending)
                return;
            _this.pending = true;
            requestAnimationFrame(_this.handlerMove);
        };
        this.handlerMove = function () {
            _this.move(_this.lastEv);
            _this.pending = false;
        };
        this.start = start;
        this.move = move;
        this.end = end;
        this.el.addEventListener('pointerdown', this.downHandler);
        this.el.addEventListener('pointerup', this.upHandler);
        this.el.addEventListener('pointercancel', this.upHandler);
    }
    Drag.prototype.dispose = function () {
        if (!this.el)
            return;
        this.el.removeEventListener('pointerdown', this.downHandler);
        this.el.removeEventListener('pointerup', this.upHandler);
        this.el.removeEventListener('pointercancel', this.upHandler);
        this.el.removeEventListener('pointermove', this.moveHandler);
        this.start = null;
        this.move = null;
        this.end = null;
        this.lastEv = null;
        this.el = null;
    };
    return Drag;
}());
export { Drag };
