// qier-player/dist/es/main/utils/freUse.js
export function adsorb(n, lower, upper) {
    if (lower === void 0) { lower = 0; }
    if (upper === void 0) { upper = 1; }
    return Math.max(Math.min(n, upper), lower);
}
export function throttle(fn, ctx) {
    var argument = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        argument[_i - 2] = arguments[_i];
    }
    var pending = false;
    var first = true;
    var args = null;
    return function () {
        var rest = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            rest[_i] = arguments[_i];
        }
        args = rest;
        if (first) {
            first = false;
            return fn.apply(ctx, args);
        }
        if (pending)
            return;
        pending = true;
        requestAnimationFrame(function () {
            fn.apply(ctx, args);
            pending = false;
        });
    };
}
function padStart(n, len, str) {
    if (len === void 0) { len = 2; }
    if (str === void 0) { str = '0'; }
    n = String(n);
    if (n.length >= 2)
        return n;
    for (var i = 0, l = len - n.length; i < l; i++) {
        n = str + n;
    }
    return n;
}
export function formatTime(s) {
    if (!Number.isFinite(s))
        return '-';
    if (s <= 0)
        return '00:00';
    s = Math.round(s);
    if (s < 60)
        return "00:" + padStart(s);
    if (s >= 60 && s < 3600) {
        var minute_1 = Math.floor(s / 60);
        var seconds_1 = Math.floor(s % 60);
        return padStart(minute_1) + ":" + padStart(seconds_1);
    }
    var hour = Math.floor(s / 3600);
    var minute = Math.floor(s / 3600 / 60);
    var seconds = Math.floor((s / 3600) % 60);
    return padStart(hour) + ":" + padStart(minute) + ":" + padStart(seconds);
}
export function repeatStr(str, t) {
    var ret = [];
    for (var i = 0; i < t; ++i) {
        ret.push(str);
    }
    return ret.join('');
}
