// qier-player/dist/es/main/utils/rect.js
import { EVENT } from '../constants';
var Rect = /** @class */ (function () {
    function Rect(el, player) {
        var _this = this;
        this.el = el;
        this.player = player;
        this.update = function () {
            _this.rect = _this.el.getBoundingClientRect();
        };
        this.rect = {};
        if (player) {
            player.on(EVENT.UPDATE_SIZE, this.update);
        }
    }
    Object.defineProperty(Rect.prototype, "width", {
        get: function () {
            this.tryUpdate(this.rect.width);
            return this.rect.width;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rect.prototype, "height", {
        get: function () {
            this.tryUpdate(this.rect.height);
            return this.rect.height;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rect.prototype, "x", {
        get: function () {
            this.tryUpdate(this.rect.left);
            return this.rect.left;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rect.prototype, "y", {
        get: function () {
            this.tryUpdate(this.rect.top);
            return this.rect.top;
        },
        enumerable: false,
        configurable: true
    });
    Rect.prototype.tryUpdate = function (v) {
        if (!v)
            this.update();
    };
    Rect.prototype.dispose = function () {
        if (this.player)
            this.player.off(EVENT.UPDATE_SIZE, this.update);
        this.el = null;
        this.rect = null;
    };
    return Rect;
}());
export { Rect };
