// qier-player/dist/es/main/modules/controller/eles/fullscreen.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { EVENT } from '../../../constants';
import { I18n, EXIT_FULL_SCREEN, FULL_SCREEN } from '../../../features/i18n';
import { Icon } from '../../../features/icons';
import { addDispose, addDisposeListener } from '../../../utils/dispose';
import { hide, show } from '../../../utils/dom';
import { DomNode } from '../../../utils/domNode';
var Fullscreen = /** @class */ (function (_super) {
    __extends(Fullscreen, _super);
    function Fullscreen() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.id = 'fullscreen';
        _this.enter = function () {
            show(_this.exitIcon);
            hide(_this.enterIcon);
            _this.tooltip.html = I18n.trans(EXIT_FULL_SCREEN);
            if (_this.mounted) {
                _this.player.updateControllerEles(_this.player.options.controller.eles.filter(function (ele) { return ele !== 'web-fullscreen'; }), 'eles');
            }
        };
        _this.exit = function () {
            hide(_this.exitIcon);
            show(_this.enterIcon);
            _this.tooltip.html = I18n.trans(FULL_SCREEN);
            if (_this.mounted) {
                _this.player.updateControllerEles(_this.player.options.controller.eles, 'eles');
            }
        };
        return _this;
    }
    Fullscreen.prototype.init = function (player, tooltip) {
        this.player = player;
        this.tooltip = tooltip;
        this.enterIcon = this.el.appendChild(Icon.enterFullscreen());
        this.exitIcon = this.el.appendChild(Icon.exitFullscreen());
        if (player.fullscreen.isActive) {
            this.enter();
        }
        else {
            this.exit();
        }
        addDispose(this, player.on(EVENT.ENTER_FULLSCREEN, this.enter));
        addDispose(this, player.on(EVENT.EXIT_FULLSCREEN, this.exit));
        addDisposeListener(this, this.el, 'click', function () { return player.fullscreen.toggle(true); });
    };
    return Fullscreen;
}(DomNode));
export var fullscreenControllerEle = function () { return new Fullscreen(); };
