// qier-player/dist/es/main/modules/controller/eles/index.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { DomNode } from '../../../utils/domNode';
import { isString } from '../../../utils/is';
import { createEle } from '../../../utils/dom';
import { Tooltip } from '../../../components/tooltip';
import { patch } from '../../../utils/patch';
import { EVENT } from '../../../constants';
import { addDispose } from '../../../utils/dispose';
import { spacerControllerEle } from './spacer';
var ControllerEle = /** @class */ (function (_super) {
    __extends(ControllerEle, _super);
    function ControllerEle(player, container, ctEles) {
        var _this = _super.call(this, container, 'div.controller_ele') || this;
        _this.player = player;
        _this.controllerEles = [];
        _this.spacer = spacerControllerEle();
        _this.initControllerEle = function (ele) {
            ele = _this.getControllerEle(ele);
            if (ele) {
                if (!ele.el)
                    ele.el = createEle('div');
                if (ele.mounted) {
                    if (ele.tooltip) {
                        ele.tooltip.resetPos();
                    }
                    return;
                }
                var tooltip = void 0;
                if (ele.tip)
                    tooltip = new Tooltip(ele.el, ele.tip);
                if (ele.init) {
                    // Do not use Tooltip when there is no tooltip parameter
                    if (ele.init.length >= 2 && !tooltip)
                        tooltip = new Tooltip(ele.el);
                    ele.init(_this.player, tooltip);
                }
                if (ele.dispose)
                    addDispose(_this, ele);
                ele.mounted = true;
                return ele;
            }
        };
        _this.onHideControllerEle = function (ele) {
            if (ele.hide)
                ele.hide();
        };
        if (ctEles) {
            var frag_1 = document.createDocumentFragment();
            ctEles.forEach(function (ele) {
                ele = _this.initControllerEle(ele);
                if (ele) {
                    frag_1.appendChild(ele.el);
                    _this.controllerEles.push(ele);
                }
            });
            _this.updateTooltipPos();
            _this.el.appendChild(frag_1);
        }
        return _this;
    }
    ControllerEle.prototype.getControllerEle = function (ele) {
        if (ele === 'spacer')
            return this.spacer;
        if (isString(ele))
            ele = this.player.getControllerEle(ele);
        // if (!item || (item.isSupport && !item.isSupport(this.player))) return;
        return ele;
    };
    ControllerEle.prototype.updateTooltipPos = function () {
        var last = this.controllerEles.length - 1;
        this.controllerEles.forEach(function (ele, i) {
            if (ele.tooltip) {
                ele.tooltip.resetPos();
                if (i === 0) {
                    ele.tooltip.setLeft();
                }
                else if (i === last) {
                    ele.tooltip.setRight();
                }
            }
        });
    };
    ControllerEle.prototype.update = function (nextEles) {
        var _this = this;
        if (!nextEles)
            return;
        var eles = [];
        var shouldUpdate = false;
        nextEles.forEach(function (ele, i) {
            ele = _this.getControllerEle(ele);
            if (!_this.controllerEles[i] || _this.controllerEles[i].id !== ele.id) {
                shouldUpdate = true;
            }
            if (ele)
                eles.push(ele);
        });
        if (shouldUpdate) {
            // Dom diff
            patch(this.controllerEles, eles, this.el, {
                mount: this.initControllerEle,
                unmount: this.onHideControllerEle,
            });
            this.controllerEles = eles;
            this.updateTooltipPos();
            this.player.emit(EVENT.CONTROLLER_ELES_UPDATE);
        }
    };
    return ControllerEle;
}(DomNode));
export { ControllerEle };
