// qier-player/dist/es/main/modules/controller/eles/play.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { DomNode } from '../../../utils/domNode';
import { show, hide } from '../../../utils/dom';
import { Icon } from '../../../features/icons';
import { addDispose, addDisposeListener } from '../../../utils/dispose';
import { EVENT } from '../../../constants';
import { I18n, PLAY, PAUSE } from '../../../features/i18n';
var Play = /** @class */ (function (_super) {
    __extends(Play, _super);
    function Play() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.id = 'play';
        _this.onPlay = function () {
            show(_this.pauseIcon);
            hide(_this.playIcon);
            _this.tooltip.html = I18n.trans(PAUSE);
        };
        _this.onPause = function () {
            show(_this.playIcon);
            hide(_this.pauseIcon);
            _this.tooltip.html = I18n.trans(PLAY);
        };
        return _this;
    }
    Play.prototype.init = function (player, tooltip) {
        this.tooltip = tooltip;
        this.playIcon = this.el.appendChild(Icon.play());
        this.pauseIcon = this.el.appendChild(Icon.pause());
        if (player.paused) {
            this.onPause();
        }
        else {
            this.onPlay();
        }
        addDispose(this, player.on(EVENT.PLAY, this.onPlay));
        addDispose(this, player.on(EVENT.PAUSE, this.onPause));
        addDisposeListener(this, this.el, 'click', player.toggle);
    };
    return Play;
}(DomNode));
export var playControllerEle = function () { return new Play(); };
