// qier-player/dist/es/main/modules/controller/eles/settings.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Popover } from '../../../components/popover';
import { Switch } from '../../../components/switch';
import { EVENT, TIME } from '../../../constants';
import { Icon } from '../../../features/icons';
import { addDispose, addDisposeListener } from '../../../utils/dispose';
import { addClass, createEle, getElementSize, removeClass } from '../../../utils/dom';
import { DomNode } from '../../../utils/domNode';
var classActive = 'controller_settings_active';
var classOptionActive = 'controller_settings_option_active';
var Settings = /** @class */ (function (_super) {
    __extends(Settings, _super);
    function Settings() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.id = 'settings';
        _this.onItemClick = function (item) { return function () {
            var _a;
            if (item.type === 'switch') {
                item.checked = !item.checked;
                (_a = item._switch) === null || _a === void 0 ? void 0 : _a.toggle(item.checked);
                if (item.change)
                    item.change(item.checked, _this.player, item);
            }
            if (item.type === 'select') {
                _this.renderOptions();
                _this.showOptionPage(item._optionEl);
            }
        }; };
        // private back = (item: ISettingItem) => () => {
        //   this.showHomePage(item._optionEl);
        // };
        _this.onOptionClick = function (item, option) { return function () {
            if (item.value !== option.value) {
                item.value = option.value;
                if (item.change)
                    item.change(option.value, _this.player, item);
            }
            // TODO 移动端不隐藏，拥有返回按钮，pc 端点击之后直接隐藏
            _this.hide();
            _this.renderHome();
            _this.showHomePage(item._optionEl);
        }; };
        _this.show = function () {
            _this.player.emit(EVENT.POPOVER_SHOW_CHANGE, _this.id);
            if (_this.popoverTimer) {
                _this.cleatTimeout();
                return;
            }
            _this.renderHome();
            _this.stuffing.style.display = 'block';
            _this.popover.show();
            _this.homeEl.style.display = 'block';
            addClass(_this.el, classActive);
        };
        _this.hide = function () {
            _this.cleatTimeout();
            _this.stuffing.style.display = 'none';
            _this.popover.hide();
            _this.homeEl.style.display = 'none';
            removeClass(_this.el, classActive);
            if (_this.currentOptionEl) {
                setTimeout(function () { return _this.showHomePage(_this.currentOptionEl); }, 200);
            }
        };
        _this.cleatTimeout = function () {
            _this.popoverTimer && clearTimeout(_this.popoverTimer);
            _this.popoverTimer = null;
        };
        _this.hideById = function (id) {
            if (id === _this.id)
                return;
            _this.hide();
        };
        return _this;
    }
    Settings.prototype.init = function (player) {
        var _this = this;
        this.player = player;
        addClass(this.el, 'controller_settings');
        this.items = player.settingItems;
        this.el.appendChild(Icon.settings());
        this.stuffing = this.el.appendChild(createEle('div.controller_settings_stuffing'));
        this.popover = new Popover(this.el, { willChange: 'width, height' });
        this.homeEl = this.popover.panelEl.appendChild(createEle('div'));
        addDisposeListener(this, this.el, 'mouseenter', function () {
            _this.show();
        });
        addDisposeListener(this, this.el, 'mouseleave', function () {
            _this.tryHide();
        });
        addDispose(this, player.on(EVENT.MOUNTED, function () { return _this.showHomePage(); }));
        addDispose(this, player.on(EVENT.POPOVER_SHOW_CHANGE, this.hideById));
        this.items.forEach(function (item) { return item.init && item.init(player, item); });
        this.renderHome();
    };
    Settings.prototype.renderHome = function () {
        var _this = this;
        this.items.forEach(function (item) {
            var el = !item._switch && !item._selectedEl && !item._optionEl ? createEle('div.controller_settings_item') : null;
            if (el) {
                el.appendChild(createEle(undefined, undefined, item.html));
                if (item.type !== 'switch')
                    el.appendChild(createEle('div.spacer'));
            }
            if (item.type === 'switch') {
                if (!item._switch)
                    item._switch = new Switch(el, item.checked);
            }
            if (item.type === 'select') {
                if (!item.options || !item.options.length)
                    return;
                if (!item._selectedEl) {
                    addClass(el, 'controller_settings_item_select');
                    item._selectedEl = el.appendChild(createEle('div', { 'data-selected': true }));
                }
                var opt = item.options.find(function (x) { return x.value === item.value; });
                if (!opt)
                    return;
                item._selectedEl.innerHTML = opt.selectedText || opt.html || '';
            }
            if (item._optionEl) {
                item._optionEl.style.display = 'none';
            }
            if (el) {
                el.addEventListener('click', _this.onItemClick(item));
                _this.homeEl.appendChild(el);
            }
        });
    };
    Settings.prototype.renderOptions = function () {
        var _this = this;
        this.items.forEach(function (item) {
            var _a;
            if (item.type !== 'select')
                return;
            if (!item._optionEl) {
                item._optionEl = _this.popover.panelEl.appendChild(createEle('div'));
                // TODO 兼容移动端时再写这部分逻辑
                // const back = item._optionEl.appendChild(
                //   createEle('div.controller_settings_item.controller_settings_item_back'),
                // );
                // back.innerHTML = item.html || '';
                // back.addEventListener('click', this.back(item));
            }
            if (!item._optionEls && item.options) {
                item._optionEls = item.options.map(function (opt) {
                    var optEl = item._optionEl.appendChild(createEle('div.controller_settings_option', undefined, opt.html));
                    optEl.addEventListener('click', _this.onOptionClick(item, opt));
                    return optEl;
                });
            }
            (_a = item._optionEls) === null || _a === void 0 ? void 0 : _a.forEach(function (optEl, i) {
                removeClass(optEl, classOptionActive);
                if (item.value === item.options[i].value) {
                    addClass(optEl, classOptionActive);
                }
            });
            item._optionEl.style.display = 'none';
        });
    };
    Settings.prototype.showHomePage = function (opt) {
        if (opt)
            opt.style.display = 'none';
        this.homeEl.style.display = '';
        var _a = getElementSize(this.homeEl), width = _a.width, height = _a.height;
        this.popover.injectPanelStyle({
            width: width + "px",
            height: height + 10 + "px", // 10 for padding
        });
    };
    Settings.prototype.showOptionPage = function (opt) {
        this.homeEl.style.display = 'none';
        opt.style.display = '';
        var _a = getElementSize(opt), width = _a.width, height = _a.height;
        this.popover.injectPanelStyle({
            width: width + "px",
            height: height + 10 + "px",
        });
        this.currentOptionEl = opt;
    };
    Settings.prototype.tryHide = function () {
        var _this = this;
        this.popoverTimer = setTimeout(function () {
            _this.hide();
        }, TIME.POPOVER_HIDE);
    };
    return Settings;
}(DomNode));
export var settingControllerEle = function () { return new Settings(); };
