// qier-player/dist/es/main/modules/controller/eles/time.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { EVENT } from '../../../constants';
import { addDispose } from '../../../utils/dispose';
import { addClass, createEle } from '../../../utils/dom';
import { DomNode } from '../../../utils/domNode';
import { formatTime } from '../../../utils/freUse';
var Time = /** @class */ (function (_super) {
    __extends(Time, _super);
    function Time() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.id = 'time';
        return _this;
    }
    Time.prototype.init = function (player) {
        var _this = this;
        addClass(this.el, 'controller_time');
        this.playedText = this.el.appendChild(createEle('span'));
        this.slashText = this.el.appendChild(createEle('span'));
        this.totalText = this.el.appendChild(createEle('span'));
        this.played = player.currentTime;
        this.total = player.duration;
        addDispose(this, player.on(EVENT.TIME_UPDATE, function () {
            _this.played = player.currentTime;
        }));
        this.slash = ' / ';
        addDispose(this, player.on(EVENT.DURATION_CHANGE, function () {
            _this.total = player.duration;
        }));
    };
    Object.defineProperty(Time.prototype, "played", {
        set: function (n) {
            this.playedText.textContent = formatTime(n);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Time.prototype, "slash", {
        set: function (s) {
            this.slashText.textContent = s;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Time.prototype, "total", {
        set: function (n) {
            this.totalText.textContent = formatTime(n);
        },
        enumerable: false,
        configurable: true
    });
    return Time;
}(DomNode));
export { Time };
export var timeControllerEle = function () { return new Time(); };
