// qier-player/dist/es/main/modules/controller/eles/volume.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { EVENT, TIME } from '../../../constants';
import { Icon } from '../../../features/icons';
import { addDispose, addDisposeListener } from '../../../utils/dispose';
import { addClass, createEle, hide, show, getEventPath } from '../../../utils/dom';
import { DomNode } from '../../../utils/domNode';
import { Drag } from '../../../utils/drag';
import { adsorb } from '../../../utils/freUse';
import { Rect } from '../../../utils/rect';
var Volume = /** @class */ (function (_super) {
    __extends(Volume, _super);
    function Volume() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.id = 'volume';
        _this.onDragStart = function (ev) {
            _this.rect.update();
            _this.onDragging(ev);
        };
        _this.onDragging = function (ev) {
            var y = _this.rect.height - (ev.y - _this.rect.y);
            var v = adsorb(y / _this.rect.height);
            _this.player.volume = v;
        };
        _this.onVolumeChange = function () {
            if (_this.player.muted) {
                _this.mute();
            }
            else {
                _this.unmute();
            }
            _this.percent.innerHTML = Math.floor(_this.player.volume * 100).toString();
            _this.percentage.style.transform = "scaleY(" + _this.player.volume + ")";
            _this.dot.style.bottom = adsorb(_this.player.volume * _this.rect.height, 0, _this.rect.height || 0) + "px";
        };
        _this.hideBarsById = function (id) {
            if (id === _this.id)
                return;
            _this.hideBars();
        };
        return _this;
    }
    Volume.prototype.init = function (player) {
        var _this = this;
        this.player = player;
        addClass(this.el, 'controller_volume');
        this.volumeIcon = this.el.appendChild(Icon.volume());
        this.mutedIcon = this.el.appendChild(Icon.muted());
        this.stuffing = this.el.appendChild(createEle('div.controller_volume_stuffing'));
        this.bars = this.el.appendChild(createEle('div.controller_volume_bars'));
        this.percent = this.bars.appendChild(createEle('span.controller_volume_percent'));
        this.barWrapper = this.bars.appendChild(createEle('div.controller_volume_bar_wrapper'));
        var volumeBar = this.barWrapper.appendChild(createEle('div.controller_volume_bar'));
        volumeBar.appendChild(createEle('div.controller_volume_bar_bg'));
        this.percentage = volumeBar.appendChild(createEle('div.controller_volume_bar_percentage'));
        this.dot = volumeBar.appendChild(createEle('div.controller_volume_bar_dot'));
        this.dot.appendChild(createEle('div.controller_volume_bar_dot_inner'));
        this.rect = new Rect(this.barWrapper, player);
        addDispose(this, player.on(EVENT.VOLUME_CHANGE, this.onVolumeChange));
        addDispose(this, new Drag(this.barWrapper, this.onDragStart, this.onDragging));
        addDisposeListener(this, this.el, 'click', function (ev) {
            var evPaths = getEventPath(ev);
            if (evPaths.includes(_this.bars) || evPaths.includes(_this.stuffing))
                return;
            player.toggleVolume();
        });
        addDisposeListener(this, this.el, 'mouseenter', function () {
            _this.showBars();
        });
        addDisposeListener(this, this.el, 'mouseleave', function () {
            _this.tryHideBars();
        });
        addDispose(this, player.on(EVENT.POPOVER_SHOW_CHANGE, this.hideBarsById));
        this.onVolumeChange();
    };
    Volume.prototype.mute = function () {
        show(this.mutedIcon);
        hide(this.volumeIcon);
    };
    Volume.prototype.unmute = function () {
        show(this.volumeIcon);
        hide(this.mutedIcon);
    };
    Volume.prototype.showBars = function () {
        this.player.emit(EVENT.POPOVER_SHOW_CHANGE, this.id);
        this.stuffing.style.display = 'block';
        this.bars.style.display = 'block';
        this.onVolumeChange();
        this.barsTimer && clearTimeout(this.barsTimer);
    };
    Volume.prototype.tryHideBars = function () {
        var _this = this;
        this.barsTimer && clearTimeout(this.barsTimer);
        this.barsTimer = setTimeout(function () {
            _this.hideBars();
        }, TIME.POPOVER_HIDE);
    };
    Volume.prototype.hideBars = function () {
        this.stuffing.style.display = 'none';
        this.bars.style.display = 'none';
    };
    return Volume;
}(DomNode));
export var volumeControllerEle = function () { return new Volume(); };
