// qier-player/dist/es/main/modules/controller/eles/web-fullscreen.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { EVENT } from '../../../constants';
import { I18n, WEB_FULL_SCREEN, WEB_EXIT_FULL_SCREEN } from '../../../features/i18n';
import { Icon } from '../../../features/icons';
import { addDispose, addDisposeListener } from '../../../utils/dispose';
import { hide, show } from '../../../utils/dom';
import { DomNode } from '../../../utils/domNode';
var WebFullscreen = /** @class */ (function (_super) {
    __extends(WebFullscreen, _super);
    function WebFullscreen() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.id = 'web-fullscreen';
        _this.enter = function () {
            show(_this.exitIcon);
            hide(_this.enterIcon);
            _this.tooltip.html = I18n.trans(WEB_EXIT_FULL_SCREEN);
        };
        _this.exit = function () {
            hide(_this.exitIcon);
            show(_this.enterIcon);
            _this.tooltip.html = I18n.trans(WEB_FULL_SCREEN);
        };
        return _this;
    }
    WebFullscreen.prototype.init = function (player, tooltip) {
        this.tooltip = tooltip;
        this.enterIcon = this.el.appendChild(Icon.webEnterFullscreen());
        this.exitIcon = this.el.appendChild(Icon.webExitFullscreen());
        if (player.webFullscreen.isActive) {
            this.enter();
        }
        else {
            this.exit();
        }
        addDispose(this, player.on(EVENT.WEB_ENTER_FULLSCREEN, this.enter));
        addDispose(this, player.on(EVENT.WEB_EXIT_FULLSCREEN, this.exit));
        addDisposeListener(this, this.el, 'click', player.webFullscreen.toggle);
    };
    return WebFullscreen;
}(DomNode));
export var webFullscreenControllerEle = function () { return new WebFullscreen(); };
