// qier-player/dist/es/main/features/fullscreen.js
import { addDisposeListener, dispose } from '../utils/dispose';
import { isFunction } from '../utils/is';
import { isIOS } from '../utils/env';
import { addClass, removeClass } from '../utils/dom';
import { EVENT } from '../constants';
var classFull = 'fullscreen';
var Fullscreen = /** @class */ (function () {
    function Fullscreen(player) {
        var _this = this;
        this.prefix = this.getPrefix();
        this.tryToggle = function () {
            _this.toggle();
        };
        this.toggle = function (isClick) {
            _this.player.clearToggleDelay(isClick);
            if (_this.isActive) {
                _this.exit();
            }
            else {
                _this.enter();
            }
        };
        this.player = player;
        addDisposeListener(this, document, this.prefix === 'ms' ? 'MSFullscreenChange' : this.prefix + "fullscreenchange", function () {
            var evt = '';
            if (_this.isActive) {
                _this.addClass();
                evt = EVENT.ENTER_FULLSCREEN;
            }
            else {
                _this.removeClass();
                evt = EVENT.EXIT_FULLSCREEN;
            }
            _this.player.emit(evt);
            _this.player.emit(EVENT.UPDATE_SIZE);
        });
        this.setTarget();
        if (this.isActive)
            this.addClass();
        // TODO 移动端适配
        this.enableDblclick();
    }
    Fullscreen.prototype.getPrefix = function () {
        if (isFunction(document.exitFullscreen))
            return '';
        var prefix = '';
        ['webkit', 'moz', 'ms'].forEach(function (p) {
            if (isFunction(document[p + "ExitFullscreen"]) ||
                isFunction(document[p + "CancelFullScreen"])) {
                prefix = p;
            }
        });
        return prefix;
    };
    Object.defineProperty(Fullscreen.prototype, "requestFullscreen", {
        get: function () {
            return (HTMLElement.prototype.requestFullscreen ||
                HTMLElement.prototype.webkitRequestFullscreen ||
                HTMLElement.prototype.mozRequestFullScreen ||
                HTMLElement.prototype.msRequestFullscreen);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Fullscreen.prototype, "exitFullscreen", {
        get: function () {
            return (HTMLDocument.prototype.exitFullscreen ||
                HTMLDocument.prototype.webkitExitFullscreen ||
                HTMLDocument.prototype.cancelFullScreen ||
                HTMLDocument.prototype.mozCancelFullScreen ||
                HTMLDocument.prototype.msExitFullscreen);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Fullscreen.prototype, "fullscreenElement", {
        get: function () {
            return (document.fullscreenElement ||
                document.mozFullScreenElement ||
                document.msFullscreenElement ||
                document.webkitFullscreenElement);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Fullscreen.prototype, "isActive", {
        get: function () {
            return this.fullscreenElement === this.target;
        },
        enumerable: false,
        configurable: true
    });
    Fullscreen.prototype.addClass = function () {
        addClass(this.player.el, classFull);
    };
    Fullscreen.prototype.removeClass = function () {
        removeClass(this.player.el, classFull);
    };
    Fullscreen.prototype.enableDblclick = function () {
        this.player.video.addEventListener('dblclick', this.tryToggle);
    };
    Fullscreen.prototype.disableDblclick = function () {
        this.player.video.removeEventListener('dblclick', this.tryToggle);
    };
    Fullscreen.prototype.setTarget = function (dom, video) {
        this.target = (dom && isIOS ? video : dom) || (isIOS ? this.player.video : this.player.el);
        if (this.isActive)
            this.enter();
    };
    Fullscreen.prototype.enter = function () {
        if (isIOS) {
            this.target.webkitEnterFullscreen();
        }
        else {
            this.requestFullscreen.call(this.target, { navigationUI: 'hide' });
            this.player.emit(EVENT.UPDATE_SIZE);
        }
    };
    Fullscreen.prototype.exit = function () {
        if (!this.isActive)
            return false;
        if (isIOS) {
            this.target.webkitExitFullscreen();
        }
        else {
            this.exitFullscreen.call(document);
            this.player.emit(EVENT.UPDATE_SIZE);
        }
        return true;
    };
    Fullscreen.prototype.dispose = function () {
        if (!this.player)
            return;
        this.player.off(EVENT.ENTER_FULLSCREEN);
        this.player.off(EVENT.EXIT_FULLSCREEN);
        this.disableDblclick();
        this.player = null;
        dispose(this);
    };
    return Fullscreen;
}());
export { Fullscreen };
