// qier-player/dist/es/main/features/i18n.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var _a;
import { isBrowser } from '../utils/env';
var data = Object.create(null);
export var I18n = {
    currentLang: '',
    defaultLang: '',
    trans: function (key, lang) {
        var _a;
        return ((_a = data[lang || this.currentLang || this.defaultLang]) === null || _a === void 0 ? void 0 : _a[key]) || key;
    },
    add: function (lang, transData) {
        data[lang.toLowerCase()] = __assign(__assign({}, data[lang.toLowerCase()]), transData);
    },
    setCurrentLang: function (lang) {
        this.currentLang =
            lang || navigator.language || navigator.browserLanguage || navigator.userLanguage;
        if (this.currentLang)
            this.currentLang = this.currentLang.toLowerCase();
    },
    setDefaultLang: function (lang) {
        this.defaultLang = (lang || '').toLowerCase();
    },
};
export var PLAY = 'Play';
export var PAUSE = 'Pause';
export var MIRRORING = 'Mirroring';
export var SPEED = 'Speed';
export var NORMAL = 'Normal';
export var WEB_FULL_SCREEN = 'Web fullscreen';
export var WEB_EXIT_FULL_SCREEN = 'Exit web fullscreen';
export var FULL_SCREEN = 'Fullscreen';
export var EXIT_FULL_SCREEN = 'Exit fullscreen';
export var LOOP = 'Loop';
export var FORWARD = 'Forward';
export var REWIND = 'Rewind';
export var SECOND = 'Second';
export var SECONDS = 'Seconds';
I18n.add('zh-cn', (_a = {},
    _a[PLAY] = '播放',
    _a[PAUSE] = '暂停',
    _a[MIRRORING] = '镜像',
    _a[SPEED] = '播放速度',
    _a[NORMAL] = '正常',
    _a[WEB_FULL_SCREEN] = '网页全屏',
    _a[WEB_EXIT_FULL_SCREEN] = '退出网页全屏',
    _a[FULL_SCREEN] = '全屏',
    _a[EXIT_FULL_SCREEN] = '退出全屏',
    _a[LOOP] = '循环播放',
    _a[FORWARD] = '快进',
    _a[REWIND] = '快退',
    _a[SECOND] = '秒',
    _a[SECONDS] = '秒',
    _a));
if (isBrowser)
    I18n.setCurrentLang();
