// qier-player/dist/es/main/features/shortcut.js
import { addDisposeListener, dispose } from '../utils/dispose';
import { createEle } from '../utils/dom';
import { isNumber, isString } from '../utils/is';
import { FORWARD, I18n, REWIND, SECOND, SECONDS } from './i18n';
import { Icon } from './icons';
var keyboardMap = {
    Escape: 27,
    Space: 32,
    ArrowLeft: 37,
    ArrowUp: 38,
    ArrowRight: 39,
    ArrowDown: 40,
};
var Shortcut = /** @class */ (function () {
    function Shortcut(player, disabled) {
        var _this = this;
        this.player = player;
        this.onKeyDown = function (ev) {
            if (ev.altKey || ev.ctrlKey || ev.metaKey || ev.shiftKey)
                return;
            // TODO 后面如果加入了弹幕到控制栏，需要再做处理。
            // const focused = document.activeElement;
            // if (focused) {
            //   const tag = focused.tagName.toLowerCase();
            //   const editable = focused.getAttribute('contenteditable');
            //   if (editableTagNames.indexOf(tag) > -1 || editable || editable === '') {
            //     return;
            //   }
            // }
            var code = ev.code || ev.keyCode || ev.which;
            if (!isNumber(code) && !isString(code))
                return;
            var handled = false;
            var handler = _this.map[keyboardMap[code]] || _this.map[code];
            if (handler) {
                handler(_this.player);
                handled = true;
                _this.player.controller.showThenFade();
            }
            if (handled) {
                ev.preventDefault();
                ev.stopPropagation();
            }
        };
        this.map = Object.create(null);
        var _a = this.player.options.shortcutOptions, seekStep = _a.seekStep, showToast = _a.showToast, toastDelay = _a.toastDelay;
        this.timeout = toastDelay || 500;
        var showVolume = function (p) {
            var icon = createEle('.toast_icon');
            var volumeIcon = Icon.volume();
            var muteIcon = Icon.muted();
            var span = createEle('span');
            span.innerHTML = Math.round(p.volume * 100) + "%";
            if (Math.floor(p.volume * 10) <= 0) {
                icon.appendChild(muteIcon);
            }
            else {
                icon.appendChild(volumeIcon);
            }
            icon.appendChild(span);
            p.toast.show(icon, 'center', _this.timeout);
        };
        this.register(keyboardMap.Escape, function (p) {
            if (!p.fullscreen.exit())
                p.webFullscreen.exit();
        });
        this.register(keyboardMap.Space, function (p) { return p.toggle(); });
        // left
        this.register(keyboardMap.ArrowLeft, function (p) {
            p.rewind();
            showToast &&
                p.toast.show(I18n.trans(REWIND) + " " + seekStep + " " + (seekStep > 1 ? I18n.trans(SECONDS) : I18n.trans(SECOND)), 'center', _this.timeout);
        });
        // top
        this.register(keyboardMap.ArrowUp, function (p) {
            p.incVolume();
            showToast && showVolume(p);
        });
        // right
        this.register(keyboardMap.ArrowRight, function (p) {
            p.forward();
            showToast &&
                p.toast.show(I18n.trans(FORWARD) + " " + seekStep + " " + (seekStep > 1 ? I18n.trans(SECONDS) : I18n.trans(SECOND)), 'center', _this.timeout);
        });
        // down
        this.register(keyboardMap.ArrowDown, function (p) {
            p.decVolume();
            showToast && showVolume(p);
        });
        if (!disabled)
            this.enable();
    }
    Shortcut.prototype.register = function (code, handler) {
        this.map[code] = handler;
    };
    Shortcut.prototype.unregister = function (code) {
        return delete this.map[code];
    };
    Shortcut.prototype.enable = function () {
        this.disable();
        addDisposeListener(this, this.player.el, 'keydown', this.onKeyDown);
    };
    Shortcut.prototype.dispose = function () {
        if (!this.map)
            return;
        dispose(this);
        this.map = null;
    };
    Shortcut.prototype.disable = function () {
        dispose(this);
    };
    return Shortcut;
}());
export { Shortcut };
