// qier-player/dist/es/main/features/web-fullscreen.js
import { EVENT } from '../constants';
import { addClass, containClass, removeClass } from '../utils/dom';
var classFull = 'web_fullscreen';
var WebFullscreen = /** @class */ (function () {
    function WebFullscreen(player) {
        var _this = this;
        this.toggle = function () {
            if (_this.isActive) {
                _this.exit();
            }
            else {
                _this.enter();
            }
        };
        this.player = player;
    }
    Object.defineProperty(WebFullscreen.prototype, "isActive", {
        get: function () {
            return containClass(this.player.el, classFull);
        },
        enumerable: false,
        configurable: true
    });
    WebFullscreen.prototype.enter = function () {
        addClass(this.player.el, classFull);
        this.player.emit(EVENT.WEB_ENTER_FULLSCREEN);
        this.player.emit(EVENT.UPDATE_SIZE);
    };
    WebFullscreen.prototype.exit = function () {
        if (!this.isActive)
            return false;
        removeClass(this.player.el, classFull);
        this.player.emit(EVENT.WEB_EXIT_FULLSCREEN);
        this.player.emit(EVENT.UPDATE_SIZE);
        return true;
    };
    WebFullscreen.prototype.dispose = function () {
        if (!this.player)
            return;
        this.player.off(EVENT.WEB_ENTER_FULLSCREEN);
        this.player.off(EVENT.WEB_EXIT_FULLSCREEN);
        this.player = null;
    };
    return WebFullscreen;
}());
export { WebFullscreen };
