// qier-player/dist/es/main/modules/toast/index.js
import { removeAllChildNode } from '../../utils/dom';
import { isHTMLElement, isString } from '../../utils/is';
import { ToastItem } from './toast-item';
var Toast = /** @class */ (function () {
    function Toast(container) {
        this.toasts = [];
        this.container = container;
    }
    Toast.prototype.show = function (html, position, timeout) {
        if (timeout === void 0) { timeout = 3000; }
        var toastItem = this.toasts.find(function (t) { return t.position === position; });
        if (toastItem) {
            clearTimeout(toastItem.timer);
            if (isHTMLElement(html)) {
                removeAllChildNode(toastItem.el);
                toastItem.el.appendChild(html);
            }
            if (isString(html))
                toastItem.el.innerHTML = html;
        }
        else {
            toastItem = new ToastItem(this.container, html, position);
            this.toasts.push(toastItem);
        }
        if (timeout > 0) {
            toastItem.timer = setTimeout(this.close.bind(this, toastItem), timeout);
        }
        return toastItem;
    };
    Toast.prototype.close = function (toastItem) {
        if (toastItem) {
            this.toasts = this.toasts.filter(function (x) { return x !== toastItem; });
            return toastItem.dispose();
        }
        this.toasts.forEach(function (item) { return item.dispose(); });
        this.toasts.length = 0;
    };
    Toast.prototype.dispose = function () {
        if (!this.toasts)
            return;
        this.close();
        this.toasts = null;
        this.container = null;
    };
    return Toast;
}());
export { Toast };
