// qier-player/dist/es/main/helper.js
import { playControllerEle } from './modules/controller/eles/play';
import { timeControllerEle } from './modules/controller/eles/time';
import { progressControllerEle } from './modules/controller/progress';
import { addDispose } from './utils/dispose';
import { EVENT } from './constants';
import { throttle } from './utils/freUse';
import { volumeControllerEle } from './modules/controller/eles/volume';
import { settingControllerEle } from './modules/controller/eles/settings';
import { MirroringSettingItem } from './setting-items/mirroring';
import { speedSettingItem } from './setting-items/speed';
import { webFullscreenControllerEle } from './modules/controller/eles/web-fullscreen';
import { fullscreenControllerEle } from './modules/controller/eles/fullscreen';
import { loopMenuItem } from './menu-items/loop';
export function setVideoAttrs(video, opts) {
    if (!opts)
        return;
    Object.keys(opts).forEach(function (k) {
        video.setAttribute(k, opts[k]);
    });
}
export function setCssVariables(el, opts) {
    var _a, _b, _c;
    var style = el.style;
    if (opts.themeColor)
        style.setProperty('--theme-color', opts.themeColor);
    if ((_a = opts.posterOptions) === null || _a === void 0 ? void 0 : _a.bgColor)
        style.setProperty('--poster-bg', opts.posterOptions.bgColor);
    if ((_b = opts.progressOptions) === null || _b === void 0 ? void 0 : _b.playedBg)
        style.setProperty('--progress-played-bg', opts.progressOptions.playedBg);
    if ((_c = opts.progressOptions) === null || _c === void 0 ? void 0 : _c.buffBg)
        style.setProperty('--progress-buff-bg', opts.progressOptions.buffBg);
}
export function registerNamedMap(player) {
    player.registerControllerEle(playControllerEle());
    player.registerControllerEle(timeControllerEle());
    player.registerControllerEle(volumeControllerEle());
    player.registerControllerEle(settingControllerEle());
    player.registerSettingItem(MirroringSettingItem());
    player.registerSettingItem(speedSettingItem());
    player.registerControllerEle(webFullscreenControllerEle());
    player.registerControllerEle(fullscreenControllerEle());
    player.registerControllerEle(progressControllerEle());
    player.registerMenuItem(loopMenuItem());
}
function mark(player, ori, event) {
    var fn = function (ev) { return player.emit(event, ev); };
    player.video.addEventListener(ori, fn);
    return { dispose: function () { return player.video.removeEventListener(ori, fn); } };
}
function markThrottle(player, ori, event, dom) {
    if (dom === void 0) { dom = player.video; }
    var fn = function (ev) { return player.emit(event, ev); };
    dom.addEventListener(ori, throttle(fn));
    return { dispose: function () { return dom.removeEventListener(ori, fn); } };
}
export function markingEvent(player) {
    var dis = function (d) { return addDispose(player, d); };
    dis(mark(player, 'play', EVENT.PLAY));
    dis(mark(player, 'pause', EVENT.PAUSE));
    dis(mark(player, 'ended', EVENT.ENDED));
    dis(mark(player, 'durationchange', EVENT.DURATION_CHANGE));
    dis(mark(player, 'waiting', EVENT.WAITING));
    dis(mark(player, 'stalled', EVENT.STALLED));
    dis(mark(player, 'canplay', EVENT.CANPLAY));
    dis(mark(player, 'loadedmetadata', EVENT.LOADED_METADATA));
    dis(mark(player, 'ratechange', EVENT.RATE_CHANGE));
    dis(mark(player, 'error', EVENT.ERROR));
    dis(mark(player, 'seeked', EVENT.SEEKED));
    dis(mark(player, 'enterpictureinpicture', EVENT.ENTER_PIP));
    dis(mark(player, 'leavepictureinpicture', EVENT.EXIT_PIP));
    dis(markThrottle(player, 'timeupdate', EVENT.TIME_UPDATE));
    dis(markThrottle(player, 'progress', EVENT.PROGRESS));
    dis(markThrottle(player, 'volumechange', EVENT.VOLUME_CHANGE));
    dis(markThrottle(player, 'resize', EVENT.UPDATE_SIZE, window));
    if (window.ResizeObserver) {
        var ro_1 = new window.ResizeObserver(throttle(function () { return player.emit(EVENT.UPDATE_SIZE); }));
        ro_1.observe(player.el);
        dis({ dispose: function () { return ro_1.disconnect(); } });
    }
    player.on(EVENT.LOADED_METADATA, function () {
        if (player.video.paused) {
            player.emit(EVENT.PAUSE);
        }
        else {
            player.emit(EVENT.PLAY);
        }
    });
}
