// qier-player/dist/es/main/options.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var defaultOptions = {
    posterOptions: {
        disabled: true,
        autoFill: true,
    },
    loadingOptions: {
        disabled: false,
        type: 'circle',
    },
    controller: {
        progress: ['progress'],
        eles: ['play', 'time', 'spacer', 'volume', 'settings', 'web-fullscreen', 'fullscreen'],
    },
    settings: ['mirroring', 'speed'],
    menus: ['loop'],
    showDefaultMenu: true,
    videoProps: {
        crossorigin: 'anonymous',
        preload: 'auto',
        playsinline: 'true',
    },
    progressOptions: {
        indicator: true,
    },
    shortcutOptions: {
        disabled: false,
        seekStep: 5,
        volumeStep: 0.1,
        showToast: true,
        toastDelay: 500,
    },
};
export function processOptions(opts) {
    var res = __assign(__assign(__assign({}, defaultOptions), opts), { posterOptions: __assign(__assign({}, defaultOptions.posterOptions), opts === null || opts === void 0 ? void 0 : opts.posterOptions), loadingOptions: __assign(__assign({}, defaultOptions.loadingOptions), opts === null || opts === void 0 ? void 0 : opts.loadingOptions), videoProps: __assign(__assign({}, defaultOptions.videoProps), opts === null || opts === void 0 ? void 0 : opts.videoProps), progressOptions: __assign(__assign({}, defaultOptions.progressOptions), opts === null || opts === void 0 ? void 0 : opts.progressOptions), shortcutOptions: __assign(__assign({}, defaultOptions.shortcutOptions), opts === null || opts === void 0 ? void 0 : opts.shortcutOptions) });
    return res;
}
export function defaultSetting(player) {
    player.volume = 0.5;
}
