// qier-player/dist/es/main/player.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { EventEmitter } from './utils/eventmitter';
import { defaultSetting, processOptions } from './options';
import { createEle, getEle, removeEle } from './utils/dom';
import { setVideoAttrs, setCssVariables, registerNamedMap, markingEvent } from './helper';
import { Controller } from './modules/controller';
import { CLASS_PREFIX, EVENT, TIME } from './constants';
import { adsorb } from './utils/freUse';
import { addDispose, dispose } from './utils/dispose';
import { isString } from './utils/is';
import { Rect } from './utils/rect';
import { WebFullscreen } from './features/web-fullscreen';
import { Fullscreen } from './features/fullscreen';
import { Poster } from './modules/poster';
import { Loading } from './modules/loading';
import { Menu } from './modules/menu';
import { Toast } from './modules/toast';
import { Shortcut } from './features/shortcut';
var Player = /** @class */ (function (_super) {
    __extends(Player, _super);
    function Player(opts) {
        var _a, _b;
        var _this = _super.call(this) || this;
        _this.prevVolume = 0.5;
        _this.controllerNameMap = Object.create(null);
        _this.settingNamedMap = Object.create(null);
        _this.menuNamedMap = Object.create(null);
        _this.toggle = function () {
            if (_this.paused) {
                _this.play();
            }
            else {
                _this.pause();
            }
        };
        _this.delayToggle = function () {
            _this.toggleDelayTimer = setTimeout(function () {
                if (!_this.toggleDelayFlag) {
                    _this.toggle();
                }
                _this.toggleDelayFlag = false;
            }, TIME.CLICK_TOGGLE_DELAY);
        };
        _this.clearToggleDelay = function (isClick) {
            _this.toggleDelayTimer && clearTimeout(_this.toggleDelayTimer);
            _this.toggleDelayTimer = null;
            if (isClick) {
                _this.toggleDelayFlag = false;
            }
            else {
                _this.toggleDelayFlag = true;
            }
        };
        _this.options = processOptions(opts);
        _this.container = getEle(_this.options.container);
        _this.el = createEle("div." + CLASS_PREFIX, { tabindex: 0 }, undefined, '');
        _this.video = createEle('video.video');
        if (_this.options.src)
            _this.options.videoProps.src = _this.options.src;
        setVideoAttrs(_this.video, _this.options.videoProps);
        setCssVariables(_this.el, _this.options);
        markingEvent(_this);
        _this.el.appendChild(_this.video);
        registerNamedMap(_this);
        _this.rect = addDispose(_this, new Rect(_this.el, _this));
        _this.webFullscreen = addDispose(_this, new WebFullscreen(_this));
        _this.fullscreen = addDispose(_this, new Fullscreen(_this));
        _this.poster = addDispose(_this, new Poster(_this.el, _this));
        _this.loading = addDispose(_this, new Loading(_this.el, _this));
        _this.toast = addDispose(_this, new Toast(_this.el));
        _this.shortcut = addDispose(_this, new Shortcut(_this, _this.options.shortcutOptions.disabled));
        _this.settingItems = _this.options.settings
            .map(function (item) {
            if (isString(item))
                return _this.settingNamedMap[item];
            return item;
        })
            .filter(Boolean);
        _this.controller = new Controller(_this, _this.el);
        _this.menu = addDispose(_this, new Menu(_this.el, _this, ((_b = (_a = _this.options.menus) === null || _a === void 0 ? void 0 : _a.map(function (item) {
            if (isString(item))
                return _this.menuNamedMap[item];
            return item;
        })) === null || _b === void 0 ? void 0 : _b.filter(Boolean)) || []));
        // To prevent click and double-click events of video elements from conflicting.
        _this.toggleDelayFlag = false;
        _this.toggleDelayTimer = null;
        _this.videoClickToggle();
        _this.emit(EVENT.AFTER_INIT);
        return _this;
    }
    Player.prototype.mount = function (container) {
        if (container)
            this.container = getEle(container) || this.container;
        if (!this.container)
            return;
        this.container.appendChild(this.el);
        defaultSetting(this);
        this.emit(EVENT.MOUNTED);
    };
    Object.defineProperty(Player.prototype, "currentTime", {
        get: function () {
            return this.video.currentTime;
        },
        set: function (n) {
            if (!this.duration)
                return;
            var diff = n - this.video.currentTime;
            if (!diff)
                return;
            this.video.currentTime = adsorb(n, 0, this.duration);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Player.prototype, "loaded", {
        get: function () {
            return this.video.readyState >= 3;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Player.prototype, "duration", {
        get: function () {
            return this.video.duration || 0;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Player.prototype, "buffered", {
        get: function () {
            return this.video.buffered;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Player.prototype, "paused", {
        get: function () {
            return this.video.paused;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Player.prototype, "ended", {
        get: function () {
            return this.video.ended;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Player.prototype, "playing", {
        get: function () {
            return !this.paused && !this.ended;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Player.prototype, "volume", {
        get: function () {
            return this.video.volume;
        },
        set: function (n) {
            this.video.volume = adsorb(n);
            if (this.muted && n > 0)
                this.muted = false;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Player.prototype, "muted", {
        get: function () {
            return this.video.muted || this.volume === 0;
        },
        set: function (v) {
            this.video.muted = v;
            if (v)
                this.volume = 0;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Player.prototype, "loop", {
        get: function () {
            return this.video.loop;
        },
        set: function (v) {
            this.video.loop = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Player.prototype, "playbackRate", {
        get: function () {
            return this.video.playbackRate;
        },
        set: function (n) {
            this.video.playbackRate = n;
        },
        enumerable: false,
        configurable: true
    });
    Player.prototype.play = function () {
        return this.video.play();
    };
    Player.prototype.pause = function () {
        this.video.pause();
    };
    Player.prototype.seek = function (seconds) {
        this.video.currentTime = adsorb(seconds, 0, this.duration);
    };
    Player.prototype.incVolume = function (v) {
        if (v === void 0) { v = this.options.shortcutOptions.volumeStep; }
        this.volume += v;
    };
    Player.prototype.decVolume = function (v) {
        if (v === void 0) { v = this.options.shortcutOptions.volumeStep; }
        this.volume -= v;
    };
    Player.prototype.forward = function (v) {
        if (v === void 0) { v = this.options.shortcutOptions.seekStep; }
        this.currentTime += v;
    };
    Player.prototype.rewind = function (v) {
        if (v === void 0) { v = this.options.shortcutOptions.seekStep; }
        this.currentTime -= v;
    };
    Player.prototype.toggleVolume = function () {
        if (this.muted) {
            this.volume = this.prevVolume || 1;
        }
        else {
            this.prevVolume = this.volume;
            this.volume = 0;
        }
    };
    Player.prototype.videoClickToggle = function () {
        this.video.addEventListener('click', this.delayToggle);
    };
    Player.prototype.cancellVideoClickToggle = function () {
        this.video.removeEventListener('click', this.delayToggle);
    };
    Player.prototype.eachBuffer = function (fn) {
        for (var l = this.buffered.length, i = l - 1; i >= 0; i--) {
            if (fn(this.buffered.start(i), this.buffered.end(i))) {
                break;
            }
        }
    };
    Player.prototype.registerControllerEle = function (ele, id) {
        this.controllerNameMap[id || ele.id] = ele;
    };
    Player.prototype.getControllerEle = function (id) {
        return this.controllerNameMap[id];
    };
    Player.prototype.registerSettingItem = function (item, id) {
        this.settingNamedMap[id || item.id] = item;
    };
    Player.prototype.getSettingItem = function (id) {
        return this.settingNamedMap[id];
    };
    Player.prototype.registerMenuItem = function (item, id) {
        this.menuNamedMap[id || item.id] = item;
    };
    Player.prototype.getMenuItem = function (id) {
        return this.menuNamedMap[id];
    };
    Player.prototype.updateControllerEles = function (eles, key) {
        this.controller.updateEles(eles, key);
    };
    Player.prototype.dispose = function () {
        if (!this.el)
            return;
        this.emit(EVENT.BEFORE_DISPOSE);
        dispose(this);
        this.removeAllListeners();
        removeEle(this.el);
        this.el = null;
        this.container = null;
        this.clearToggleDelay();
        this.emit(EVENT.AFTER_DISPOSE);
    };
    return Player;
}(EventEmitter));
export { Player };
