// qier-player/dist/es/main/modules/controller/progress/index.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { EVENT } from '../../../constants';
import { addDispose, addDisposeListener } from '../../../utils/dispose';
import { addClass, createEle } from '../../../utils/dom';
import { DomNode } from '../../../utils/domNode';
import { Drag } from '../../../utils/drag';
import { adsorb, throttle } from '../../../utils/freUse';
import { Rect } from '../../../utils/rect';
import { Thumbnail } from './thumbnail';
var Progress = /** @class */ (function (_super) {
    __extends(Progress, _super);
    function Progress() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.id = 'progress';
        _this.dragging = false;
        _this.onDragStart = function (ev) {
            _this.dragging = true;
            _this.rect.update();
            _this.onDragging(ev);
        };
        _this.onDragging = function (ev) {
            var x = ev.pageX - _this.rect.x;
            _this.setPlayedBarLength(x / _this.rect.width);
            _this.updateHoverElement(ev.pageX);
        };
        _this.onDragEnd = function (ev) {
            _this.dragging = false;
            _this.player.seek(_this.getCurrentTime(ev.pageX));
        };
        _this.updatePlayedBar = function () {
            if (_this.dragging)
                return;
            _this.setPlayedBarLength(_this.player.currentTime / _this.player.duration);
        };
        _this.updateBuffBar = function () {
            var buffLen = _this.player.buffered.length;
            if (!buffLen)
                return _this.setBuffBarLength(0);
            var curTime = _this.player.currentTime;
            var percentage = 0;
            _this.player.eachBuffer(function (start, end) {
                if (start <= curTime && end >= curTime) {
                    percentage = end / _this.player.duration;
                    return true;
                }
            });
            _this.setBuffBarLength(percentage);
        };
        return _this;
    }
    Progress.prototype.init = function (player) {
        var _this = this;
        var _a, _b;
        this.player = player;
        addClass(this.el, 'progress');
        this.bars = this.el.appendChild(createEle('div.progress_bars'));
        this.buffBar = this.bars.appendChild(createEle('div.progress_buff'));
        this.playedBar = this.bars.appendChild(createEle('div.progress_played'));
        this.dot = this.el.appendChild(createEle('div.progress_dot'));
        this.dot.appendChild(((_a = player.options.progressOptions) === null || _a === void 0 ? void 0 : _a.dot) || createEle('div.progress_dot_inner'));
        this.indicator = this.el.appendChild(createEle('div.progress_indicator'));
        if ((_b = player.options.progressOptions) === null || _b === void 0 ? void 0 : _b.indicator) {
            this.indicator.appendChild(createEle('div.progress_indicator_inner'));
        }
        this.rect = addDispose(this, new Rect(this.bars, player));
        this.thumbnail = addDispose(this, new Thumbnail(this.el, player.options.thumbnail));
        addDispose(this, new Drag(this.el, this.onDragStart, this.onDragging, this.onDragEnd));
        addDispose(this, player.on(EVENT.TIME_UPDATE, this.updatePlayedBar));
        addDispose(this, player.on(EVENT.PROGRESS, this.updateBuffBar));
        addDispose(this, player.on(EVENT.UPDATE_SIZE, function () {
            if (!player.playing)
                _this.resetPlayedBar();
        }));
        addDisposeListener(this, this.el, 'mousemove', throttle(function (ev) { return _this.updateHoverElement(ev.pageX); }));
    };
    Progress.prototype.resetPlayedBar = function () {
        this.setPlayedBarLength(this.player.currentTime / this.player.duration);
    };
    Progress.prototype.setPlayedBarLength = function (percentage) {
        this.playedBar.style.transform = "scaleX(" + adsorb(percentage) + ")";
        this.dot.style.left = adsorb(percentage * this.rect.width, 0, this.rect.width || 0) + "px";
    };
    Progress.prototype.setBuffBarLength = function (percentage) {
        this.buffBar.style.transform = "scaleX(" + adsorb(percentage) + ")";
    };
    Progress.prototype.updateHoverElement = function (x) {
        this.thumbnail.updatePos(this.getCurrentTime(x), x - this.rect.x, this.rect.width);
        if (this.indicator) {
            this.indicator.style.left = adsorb(x - this.rect.x, 0, this.rect.width || 0) + "px";
        }
    };
    Progress.prototype.getCurrentTime = function (x) {
        return adsorb((x - this.rect.x) / this.rect.width) * this.player.duration;
    };
    return Progress;
}(DomNode));
export var progressControllerEle = function () { return new Progress(); };
