// qier-player/dist/es/main/modules/controller/progress/thumbnail.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { addDispose } from '../../../utils/dispose';
import { createEle } from '../../../utils/dom';
import { DomNode } from '../../../utils/domNode';
import { adsorb, formatTime } from '../../../utils/freUse';
import { Rect } from '../../../utils/rect';
var defaultOpts = {
    startSecond: 0,
    gapSecond: 5,
    col: 5,
    row: 5,
    width: 160,
    height: 90,
    images: [],
};
var Thumbnail = /** @class */ (function (_super) {
    __extends(Thumbnail, _super);
    function Thumbnail(container, opts) {
        var _this = _super.call(this, container, 'div.progress_thumb') || this;
        _this.opts = defaultOpts;
        _this.updateOptions(opts);
        _this.timeTagEl = _this.el.appendChild(createEle('div.progress_thumb_time'));
        _this.timeTagEl.textContent = '0:00';
        _this.rect = addDispose(_this, new Rect(_this.el));
        return _this;
    }
    Thumbnail.prototype.getCurrentThumb = function (second) {
        var preImgIndex = Math.floor((second - this.opts.startSecond) / ((this.onePreImgNums || 1) * this.opts.gapSecond));
        var imgIndex = Math.floor(((second - this.opts.startSecond) % ((this.onePreImgNums || 1) * this.opts.gapSecond)) / this.opts.gapSecond);
        var url = this.opts.images[preImgIndex];
        if (!url)
            return;
        var x = ((imgIndex % this.opts.col) % this.opts.col) * this.opts.width;
        var y = (Math.floor(imgIndex / this.opts.col) % this.opts.row) * this.opts.height;
        return { url: url, x: x, y: y };
    };
    Thumbnail.prototype.updateOptions = function (opts) {
        Object.assign(this.opts, opts);
        if (!this.opts.images.length)
            return;
        if (!this.preImgEl) {
            this.preImgEl = this.el.appendChild(createEle('div.progress_thumb_img'));
        }
        this.preImgEl.style.cssText += ";width:" + this.opts.width + "px;height:" + this.opts.height + "px";
        this.onePreImgNums = this.opts.row * this.opts.col;
    };
    Thumbnail.prototype.updatePos = function (seconds, x, maxX) {
        this.timeTagEl.textContent = formatTime(seconds);
        if (this.preImgEl) {
            var thumb = this.getCurrentThumb(seconds);
            if (thumb) {
                this.preImgEl.style.cssText += ";background-image:url(" + thumb.url + ");background-position:-" + thumb.x + "px -" + thumb.y + "px";
            }
        }
        else {
            this.rect.update();
        }
        var half = this.rect.width / 2;
        this.injectStyle({
            left: adsorb(x - half, 0, Math.max(0, maxX - 2 * half)) + "px",
        });
    };
    return Thumbnail;
}(DomNode));
export { Thumbnail };
